components/cqrs/event_processor_group.go

```diff
@@ -0,0 +1,260 @@
+//
+// Handlers group needs to be unique within the EventProcessor instance.
+func (p *EventGroupProcessor) AddHandlersGroup(handlerName string, handlers ...GroupEventHandler) error {
+	if len(handlers) == 0 {
+		return errors.New("no handlers provided")
+	}
+	if _, ok := p.groupEventHandlers[handlerName]; ok {
+		return fmt.Errorf("event handler group '%s' already exists", handlerName)
+	}
+
+	p.groupEventHandlers[handlerName] = handlers
+
+	return nil
+}
+
+// AddHandlersToRouter adds the EventProcessor's handlers to the given router.
+// It should be called only once per EventProcessor instance.
+func (p EventGroupProcessor) AddHandlersToRouter(r *message.Router) error {
+	if len(p.groupEventHandlers) == 0 {
+		return errors.New("EventProcessor has no handlers, did you call AddHandlersGroup?")
+	}
+
+	for groupName := range p.groupEventHandlers {
+		handlersGroup := p.groupEventHandlers[groupName]
+
+		for i, handler := range handlersGroup {
+			if err := validateEvent(handler.NewEvent()); err != nil {
+				return fmt.Errorf(
+					"invalid event for handler %T (num %d) in group %s: %w",
+					handler,
+					i,
+					groupName,
+					err,
+				)
+			}
+		}
+
+		if p.config.GenerateSubscribeTopic == nil {
+			return errors.New("missing GenerateSubscribeTopic config option")
+		}
+
+		topicName, err := p.config.GenerateSubscribeTopic(EventGroupProcessorGenerateSubscribeTopicParams{
+			EventGroupName:     groupName,
+			EventGroupHandlers: handlersGroup,
+		})
+		if err != nil {
+			return errors.Wrapf(err, "cannot generate topic name for handler group %s", groupName)
+		}
+
+		logger := p.config.Logger.With(watermill.LogFields{
+			"event_handler_group_name": groupName,
+			"topic":                    topicName,
+		})
+
+		handlerFunc, err := p.routerHandlerGroupFunc(handlersGroup, groupName, logger)
+		if err != nil {
+			return err
+		}
+
+		subscriber, err := p.config.SubscriberConstructor(EventGroupProcessorSubscriberConstructorParams{
+			EventGroupName:     groupName,
+			EventGroupHandlers: handlersGroup,
+		})
+		if err != nil {
+			return errors.Wrap(err, "cannot create subscriber for event processor")
+		}
+
+		if err := addHandlerToRouter(p.config.Logger, r, groupName, topicName, handlerFunc, subscriber); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p EventGroupProcessor) routerHandlerGroupFunc(handlers []GroupEventHandler, groupName string, logger watermill.LoggerAdapter) (message.NoPublishHandlerFunc, error) {
+	return func(msg *message.Message) error {
+		messageEventName := p.config.Marshaler.NameFromMessage(msg)
+
+		for _, handler := range handlers {
+			initEvent := handler.NewEvent()
+			expectedEventName := p.config.Marshaler.Name(initEvent)
+
+			event := handler.NewEvent()
+
+			if messageEventName != expectedEventName {
+				logger.Trace("Received different event type than expected, ignoring", watermill.LogFields{
```

Comment:
This is expected, right? We shouldn't raise this in any way?

Reply:
It's handled after the loop (this loop is looping over handlers in group - so the next handler may be the handler that should handle it)